monkey/agent_plugins/exploiters/hadoop/src/hadoop_command_builder.py

```diff
@@ -1,71 +1,82 @@
+    agent_destination_path: PurePath,
     agent_download_url: str,
-    agent_otp_environment_variable: str,
-    otp: str,
+    agent_command_builder_factory: IAgentCommandBuilderFactory,
 ) -> str:
-    monkey_cmd = build_monkey_commandline(agent_id, servers, current_depth + 1)
-
     if target_host.operating_system in [
         OperatingSystem.WINDOWS,
         None,  # if unknown OS, default to Windows (based on
         # https://www.shodan.io/search/facet?query=hadoop&facet=os)
     ]:
-        base_command = HADOOP_WINDOWS_COMMAND_TEMPLATE
+        return _build_windows_hadoop_command(
+            PureWindowsPath(agent_destination_path),
+            agent_download_url,
+            agent_command_builder_factory,
+        )
     else:
-        base_command = HADOOP_LINUX_COMMAND_TEMPLATE
+        return _build_linux_hadoop_command(
+            PurePosixPath(agent_destination_path), agent_download_url, agent_command_builder_factory
+        )
+
+
+def _build_windows_hadoop_command(
+    agent_destination_path: PureWindowsPath,
+    agent_download_url: str,
+    agent_command_builder_factory: IAgentCommandBuilderFactory,
+) -> str:
+    download_options = WindowsDownloadOptions(
+        agent_destination_path=agent_destination_path,
+        download_method=WindowsDownloadMethod.WEB_REQUEST,
+        download_url=agent_download_url,
+    )
+
+    run_options = WindowsRunOptions(
+        agent_destination_path=agent_destination_path,
+        dropper_execution_mode=DropperExecutionMode.NONE,
+        shell=WindowsShell.POWERSHELL,
+    )
+
+    agent_command_builder = agent_command_builder_factory.create_windows_agent_command_builder()
+    agent_command_builder.build_download_command(download_options)
+    agent_command_builder.build_run_command(run_options)
+    return agent_command_builder.get_command()
+
+
+def _build_linux_hadoop_command(
+    agent_destination_path: PurePosixPath,
+    agent_download_url: str,
+    agent_command_builder_factory: IAgentCommandBuilderFactory,
+) -> str:
+    download_options = LinuxDownloadOptions(
+        agent_destination_path=agent_destination_path,
+        download_method=LinuxDownloadMethod.WGET,
+        download_url=agent_download_url,
+    )
+
+    run_options = LinuxRunOptions(
+        agent_destination_path=agent_destination_path,
+        dropper_execution_mode=DropperExecutionMode.NONE,
+    )
 
-    return base_command % {
-        "monkey_path": agent_destination_path,
-        "http_path": agent_download_url,
-        "monkey_type": MONKEY_ARG,
-        "parameters": monkey_cmd,
-        "agent_otp_environment_variable": agent_otp_environment_variable,
-        "agent_otp": otp,
-    }
+    agent_command_builder = agent_command_builder_factory.create_linux_agent_command_builder()
+    agent_command_builder.build_download_command(download_options)
+    agent_command_builder.build_run_command(run_options)
+    return agent_command_builder.get_command()
```

Comment:
Do these two functions need to be coupled to the factory?

Reply:
We can pass the concrete builders but I don't see a point why would we do that.

Reply:
Minimize coupling. Components should depend on only what they need.